python/cudf/cudf/core/buffer.py

```diff
@@ -1,218 +1,61 @@
-        iface = header["cuda_array_interface"]
-        if (
-            len(iface["shape"]) == 0 or 0 in iface["shape"]
-        ):  # The array is empty
-            arr = rmm.device_array(
-                iface["shape"], dtype=np.dtype(iface["typestr"])
-            )
+class Buffer:
+    def __init__(self, ptr, size=None, owner=None):
+        if not ptr:
+            self.ptr = 0
+            self.size = 0
         else:
-            # Updating the data pointer to the frame data.
-            iface["data"] = frames[0].__cuda_array_interface__["data"]
-
-            # We need an object that exposes __cuda_array_interface__
-            _dummy_iface = types.SimpleNamespace()
-            _dummy_iface.__cuda_array_interface__ = iface
-
-            # Allocating a new RMM CUDA array with shape and dtype as specified
-            # in `iface` and copy the frame data to it, which makes the memory
-            # survive until the new Buffer goes out of scope.
-            # TODO: when DASK supports RMM, we can simply return it as is.
-            arr, _ = rmm.auto_device(_dummy_iface)
-        return Buffer(arr)
+            self.ptr = ptr
+            self.size = size
+        self._owner = owner
 
     def __reduce__(self):
-        cpumem = self.to_array()
-        # Note: pickled Buffer only stores *size* element.
-        return type(self), (cpumem,)
+        return self.__class__.from_array_like, (self.to_host_array(),)
 
-    def __sizeof__(self):
-        return int(self.mem.alloc_size)
+    def to_host_array(self):
+        return rmm.device_array_from_ptr(
+            self.ptr, nelem=self.size, dtype="int8"
+        ).copy_to_host()
 
-    def __getitem__(self, arg):
-        if isinstance(arg, slice):
-            sliced = self.mem[arg]
-            buf = Buffer(sliced)
-            buf.dtype = self.dtype  # for np.datetime64 support
-            return buf
-        elif isinstance(arg, (int, np.integer)):
-            arg = utils.normalize_index(int(arg), self.size)
-            item = self.mem[arg]
-            if isinstance(item, str):
-                return item
-            # the dtype argument is necessary for datetime64 support
-            # because currently we can't pass datetime64 types into
-            # cuda dev arrays, so the type of the cuda dev array is
-            # an i64, and we view it as the dtype on the buffer
-            return item.view(self.dtype)
+    @classmethod
+    def from_array_like(cls, data):
+        if hasattr(data, "__cuda_array_interface__"):
+            ptr, size = _buffer_data_from_array_interface(
+                data.__cuda_array_interface__
+            )
+            return Buffer(ptr, size, owner=data)
+        elif hasattr(data, "__array_interface__"):
+            ptr, size = _buffer_data_from_array_interface(
+                data.__array_interface__
+            )
+            dbuf = DeviceBuffer(ptr, size)
```

Comment:
Seems like most cases where a `DeviceBuffer` is constructed keyword arguments are used already. Might be nice to do that here too.

```suggestion
            dbuf = DeviceBuffer(ptr=ptr, size=size)
```